backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsync.java

```diff
@@ -0,0 +1,197 @@
+                extractedData.getLastName().getValue(),
+                extractedData.getSuffix().getValue(),
+                parseUserShortDate(extractedData.getDateOfBirth().getValue()),
+                address,
+                country,
+                parsePersonRole(extractedData.getRole().getValue(), false),
+                List.of(extractedData.getEmail().getValue()),
+                convertRaceToDatabaseValue(extractedData.getRace().getValue()),
+                convertEthnicityToDatabaseValue(extractedData.getEthnicity().getValue()),
+                null, // tribalAffiliation
+                convertSexToDatabaseValue(extractedData.getBiologicalSex().getValue()),
+                parseYesNo(extractedData.getResidentCongregateSetting().getValue()),
+                parseYesNo(extractedData.getEmployedInHealthcare().getValue()),
+                null, // preferredLanguage
+                null // testResultDeliveryPreference
+                );
+
+        if (!patientsList.contains(newPatient)) {
+          // collect phone numbers and associate them with the patient
+          // then add to phone numbers list and set primary phone, if exists
+          List<PhoneNumber> newPhoneNumbers =
+              _personService.assignPhoneNumbersToPatient(
+                  newPatient,
+                  List.of(
+                      new PhoneNumber(
+                          parsePhoneType(extractedData.getPhoneNumberType().getValue()),
+                          extractedData.getPhoneNumber().getValue())));
+          phoneNumbersList.addAll(newPhoneNumbers);
+          newPhoneNumbers.stream().findFirst().ifPresent(newPatient::setPrimaryPhone);
+
+          patientsList.add(newPatient);
+        }
+      } catch (IllegalArgumentException e) {
+        sendErrorEmail(uploaderEmail, simplereportUrl, currentOrganization);
+        String errorMessage = "Error uploading patient roster";
+        log.error(
+            errorMessage
+                + " for organization "
+                + currentOrganization.getExternalId()
+                + " and facility "
+                + facilityId);
+        throw new IllegalArgumentException(errorMessage);
+      }
+    }
+
+    try {
+      _personService.addPatientsAndPhoneNumbers(patientsList, phoneNumbersList);
+      log.info("CSV patient upload completed for {}", currentOrganization.getOrganizationName());
+      sendSuccessEmail(uploaderEmail, patientsUrl, currentOrganization);
+    } catch (IllegalArgumentException | OptimisticLockingFailureException e) {
+      sendErrorEmail(uploaderEmail, simplereportUrl, currentOrganization);
+      String errorMessage = "Error saving patient roster";
+      log.error(
+          errorMessage
+              + " for organization "
+              + currentOrganization.getExternalId()
+              + " and facility "
+              + facilityId);
+      throw new IllegalArgumentException(errorMessage);
+    }
+  }
+
+  private void sendErrorEmail(
```

Comment:
There's a lot of duplicate code between this and the `sendSuccessEmail` method - I wonder if you could consolidate down to a single `sendEmail` method and pass a boolean for success/failure or something like that. Up to you though, this way may be more readable.

Reply:
✅